org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/secondgeneration/SecondGenerationHandler.java

```diff
@@ -0,0 +1,513 @@
+import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link SecondGenerationHandler} is responsible for handling commands, which are
+ * sent to one of the channels, regarded to second generation part of the binding.
+ *
+ * @author Christian Schneider - Initial contribution
+ * @author Christoph Weitkamp - Incorporated new QuantityType (Units of Measurement)
+ * @author Örjan Backsell - Redesigned regarding Piko1020, Piko New Generation models
+ */
+// @NonNullByDefault
+public class SecondGenerationHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(SecondGenerationHandler.class);
+
+    // @Nullable
+    private SecondGenerationConfiguration config;
+
+    @SuppressWarnings("unused")
+    // @Nullable
+    private SecondGenerationBindingConstants configurationConfig;
+
+    private final List<SecondGenerationChannelConfig> channelConfigs = new ArrayList<>();
+    private final List<SecondGenerationChannelConfig> channelConfigsExt = new ArrayList<>();
+    private final List<SecondGenerationChannelConfig> channelConfigsExtExt = new ArrayList<>();
+
+    public SecondGenerationHandler(Thing thing) {
+        super(thing);
+    }
+
+    // @SuppressWarnings("null")
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String url = config.url.toString();
+        String username = config.username;
+        String password = config.password;
+        String dxsIdConfiguration = "";
+        String valueConfiguration = "";
+
+        if (command instanceof RefreshType) {
+            logger.debug("Handle command for {} on channel {}: {}", thing.getUID(), channelUID, command);
+        }
+
+        if (command == RefreshType.REFRESH) {
+            return;
+        }
+
+        if (channelUID.getId().equals("chargeTimeEnd")) {
+            dxsIdConfiguration = "33556236";
+            valueConfiguration = command.toString();
+
+            try {
+                SecondGenerationConfigurationHandler.configurationHandler(url, username, password, dxsIdConfiguration,
+                        valueConfiguration);
+            } catch (Exception e) {
+                logger.error("Unexpected error in KostalConfigurationHandler.configurationHandler, chargeTimeEnd !", e);
```

Comment:
The error logging level should be reserved for issues catastrophic enough to threaten the operation of openHab itself. In error in a binding shouldn't qualify for this. Please read the [openHab coding guidelines](https://www.openhab.org/docs/developer/guidelines.html#f-logging) for expected use of logging in bindings.

My general rule for logging levels is:
* trace - used for normal expected execution paths
* debug - used for unexepected execution paths, but well within normal operation
* info - used for notable points in an execution path, like a milestone. (in openhab we try reserve this logging level for the core, so bindings should rarely ever call this.)
* warn - used for notable unexpected execution paths that a regular user (not just a developer) should be notified of. Warnings should be used to indicate that something not-normal occurred and user intervention is required to resolve. Warnings do not indicate a failure to operate merely an abnormal condition of operation that can still be handled by the binding. Failures in binding operation should be indicated by changing the thing status to offline.
* error - used to indicate catastrophic program failure. This should be used to indicate a catastrophic failure in openhab's ability to operate. A failure in a binding would never cause openhab as a whole to fail so a failure in a bindings should never log an error. Instead that failure should be indicated by changing the thing status.